### PR TITLE
Fixed behavior of `insertOrUpdate` on tables with auto_increment option

### PR DESCRIPTION
**Pull request by jkugiya**
#2784

### Problem
The part of on duplicate key update that specifies the update target has included an auto_increment column since 3.4.1, so the behavior of upsert on tables with unique columns has changed.

#### Example

```scala
      case class AnimalRow(id: Int, name: String, location: String)
      class Animals(tag: Tag) extends Table[AnimalRow](tag, "animals") {
        def id = column[Int]("id", O.PrimaryKey, O.AutoInc)
        def name = column[String]("name", O.Unique, O.Length(100))
        def location = column[String]("location")
        def * = (id, name, location) <> (AnimalRow.tupled, AnimalRow.unapply)
      }
      val animals = TableQuery[Animals]
      for {
        _ <- animals.schema.create
        _ <- animals.insertOrUpdate(AnimalRow(0, "tiger", "NY"))
        tigerAfterInsert <- animals.filter(_.name === "tiger").result.head
        _ <- animals.insertOrUpdate(AnimalRow(0, "tiger", "CA"))
        tigerAfterUpsertByUniqueName <- animals.filter(_.name === "tiger").result.head
      } yield {
        tigerAfterInsert.id shouldBe tigerAfterUpsertByUniqueName.id
        tigerAfterUpsertByUniqueName.location shouldBe "CA"
      }
```

In this example, the following SQL was issued in slick3.3.3.(The second statement is updated because "tiger" is duplicated)
```sql
insert ignore into `animals` (`id`,`name`,`location`)  values (?,?,?) on duplicate key update `name`=VALUES(`name`), `location`=VALUES(`location`)
```

The following SQL is issued in slick3.4.1. (The auto_increment column is not considered a duplicate, so the second statement is also inserted)

```sql
insert into `animals` (`id`,`name`,`location`)  values (?,?,?) on duplicate key update `id`=VALUES(`id`), `name`=VALUES(`name`), `location`=VALUES(`location`)
```

### Revision details
The following SQL is now issued.

```sql
insert into `animals` (`id`,`name`,`location`)  values (?,?,?) on duplicate key update name=VALUES(name), location=VALUES(location)
```

Since only auto_increment columns are excluded from duplicate key checks, #1627 will not recur

```scala
      case class V(id: Int, value: String)
      class T2(tag: Tag) extends Table[V](tag, "mytable2") {
        def id = column[Int]("id")
        def value = column[String]("value", O.Length(20))
        def * = (id, value) <> ((V.apply _).tupled, V.unapply)
        def pk = primaryKey("pk", (id, value))
      }
      val ts2 = TableQuery[T2]
      println(ts2.insertOrUpdate(V(0, "c")).statements.mkString)
      // insert into `mytable2` (`id`,`value`)  values (?,?) on duplicate key update id=VALUES(id), value=VALUES(value)
```

### CODE REVIEW

File: slick-testkit/src/test/scala/slick/test/lifted/MySQLDMLTest.scala [hunk 1]
```diff
@@ -1,25 +0,0 @@
...
-    class T2(tag: Tag) extends Table[V](tag, "mytable2") {
-      def id = column[String]("id", O.PrimaryKey)
-      def value = column[String]("value")
-      def * = (id, value).<>((V.apply _).tupled, V.unapply)
-    }
-    assertTrue("generates query with update action if insert primary key only",
-      TableQuery[T].insertOrUpdate(1).statements.mkString.endsWith("on duplicate key update `id`=VALUES(`id`)"))
-    assertTrue("generates query with update action if insert with value",
-      TableQuery[T2].insertOrUpdate(V("test", "test-value")).statements.mkString.endsWith("on duplicate key update `id`=VALUES(`id`), `value`=VALUES(`value`)"))
-  }
```
Comment (jkugiya) [inline slick-testkit/src/test/scala/slick/test/lifted/MySQLDMLTest.scala:24]: I have decided that the test code is not worth much now.

**Comment by jkugiya** [comment]
I will close it once because it seems that I have misunderstood the issue.

**Comment by nafg** [comment]
@mdedetrich can you fix

**Comment by mdedetrich** [comment]
> @mdedetrich can you fix

@jkugiya What do you want me to do here, review the PR/confirm that it works?

**Comment by jkugiya** [comment]
I had reopened this because my misunderstand was solved.

**Comment by jkugiya** [comment]
> review the PR/confirm that it works 

Please review it.

**Comment by mdedetrich** [comment]
Okay I will do it at the end of the day

**Comment by nafg** [comment]
No, I mean the build is broken by your pr, check the failing job. Otherwise I'll get to it

**Comment by mdedetrich** [comment]
> No, I mean the build is broken by your pr, check the failing job. Otherwise I'll get to it

Ah I see thanks, I will quickly look into this. I am not sure how my PR should have broken this specifically (rather than just generally), also not too familiar with the PR compat report but ill dig into it.

**Comment by nafg** [comment]
@mergifyio squash

**Review by nafg — APPROVED** [review]

**Comment by nafg** [comment]
Thanks!

**Comment by nafg** [comment]
@Mergifyio backport 3.4.x

**Comment by jkugiya** [comment]
@nafg 
When you squash, doesn't the cotribute disappear? Why is my modification your commit?

8665e0e151e07c9a5b39db8899541c9222f11577

**Comment by nafg** [comment]
@jkugiya really sorry about that. I can't remember but it's possible it was a few fragmented commits and I didn't realize Mergify would overwrite the author field. When you use git rebase to squash it keeps the author and only changes the committer.

**Pull request merged**